control-plane-operator/controllers/hostedcontrolplane/dnsoperator/dnsoperator.go

```diff
@@ -0,0 +1,190 @@
+package dnsoperator
+
+import (
+	hyperv1 "github.com/openshift/hypershift/api/v1alpha1"
+	"github.com/openshift/hypershift/control-plane-operator/controllers/hostedcontrolplane/kas"
+	"github.com/openshift/hypershift/support/config"
+	"github.com/openshift/hypershift/support/util"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilpointer "k8s.io/utils/pointer"
+)
+
+const (
+	// dnsOperatorContainerName is the name of the operator container.
+	dnsOperatorContainerName = "dns-operator"
+)
+
+// Images stores the image pullspecs for the images that the DNS operator
+// deployment references.
+type Images struct {
+	DNSOperator   string
+	CoreDNS       string
+	KubeRBACProxy string
+	CLI           string
+}
+
+// Params stores the parameters that are required to configure the DNS operator
+// deployment on HyperShift, as well as some additional information about the
+// deployment.
+type Params struct {
+	// ReleaseVersion is the HyperShift release version.
+	ReleaseVersion string
+	// AvailabilityProberImage is the image for the prober, which runs on
+	// the management cluster and probes the DNS operator.
+	AvailabilityProberImage string
+	// Images has the image pullspecs that the DNS operator deployment
+	// references.
+	Images Images
+	// OwnerRef is a reference to the HyperShift control plane.
+	OwnerRef config.OwnerRef
+	// DeploymentConfig has information about the DNS operator deployment.
+	DeploymentConfig config.DeploymentConfig
+}
+
+// NewParams creates a new Params object for a DNS operator deployment.
+func NewParams(hcp *hyperv1.HostedControlPlane, version string, images map[string]string) Params {
+	p := Params{
+		Images: Images{
+			DNSOperator:   images["cluster-dns-operator"],
+			CoreDNS:       images["coredns"],
+			KubeRBACProxy: images["kube-rbac-proxy"],
+			CLI:           images["cli"],
+		},
+		ReleaseVersion:          version,
+		AvailabilityProberImage: images[util.AvailabilityProberImageName],
+		OwnerRef:                config.OwnerRefFrom(hcp),
+	}
+
+	p.DeploymentConfig.AdditionalAnnotations = map[string]string{
+		"target.workload.openshift.io/management": `{"effect": "PreferredDuringScheduling"}`,
+	}
+	p.DeploymentConfig.AdditionalLabels = map[string]string{
+		"name": "dns-operator",
+	}
+	p.DeploymentConfig.Scheduling.PriorityClass = config.DefaultPriorityClass
+	p.DeploymentConfig.SetRestartAnnotation(hcp.ObjectMeta)
+	p.DeploymentConfig.SetDefaults(hcp, nil, utilpointer.IntPtr(1))
+	p.DeploymentConfig.SetDefaultSecurityContext = false
```

Comment:
why false? this needs to come from r.SetDefaultSecurityContext so is set based on SCC capability.

Reply:
I'm not familiar with HyperShift, and `SetDefaultSecurityContext` is undocumented, so if you say `SetDefaultSecurityContext = true` would be better, I'll make that change.

Reply:
Now creation of the dns-operator pod is failing
```yaml
  - lastTransitionTime: "2022-08-23T15:14:03Z"
    lastUpdateTime: "2022-08-23T15:14:03Z"
    message: 'pods "dns-operator-69d5fc8bf8-" is forbidden: unable to validate against
      any security context constraint: [provider "anyuid": Forbidden: not usable by
      user or serviceaccount, spec.initContainers[0].securityContext.runAsUser: Invalid
      value: 1001: must be in the ranges: [1000770000, 1000779999], spec.initContainers[1].securityContext.runAsUser:
      Invalid value: 1001: must be in the ranges: [1000770000, 1000779999], spec.containers[0].securityContext.runAsUser:
      Invalid value: 1001: must be in the ranges: [1000770000, 1000779999], provider
      "restricted": Forbidden: not usable by user or serviceaccount, provider "nonroot-v2":
      Forbidden: not usable by user or serviceaccount, provider "nonroot": Forbidden:
      not usable by user or serviceaccount, provider "hostmount-anyuid": Forbidden:
      not usable by user or serviceaccount, provider "hostnetwork-v2": Forbidden:
      not usable by user or serviceaccount, provider "hostnetwork": Forbidden: not
      usable by user or serviceaccount, provider "hostaccess": Forbidden: not usable
      by user or serviceaccount, provider "node-exporter": Forbidden: not usable by
      user or serviceaccount, provider "privileged": Forbidden: not usable by user
      or serviceaccount]'
    reason: FailedCreate
    status: "True"
    type: ReplicaFailure
```
https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_hypershift/1537/pull-ci-openshift-hypershift-main-e2e-aws-nested/1562088835695775744/artifacts/e2e-aws-nested/run-e2e/artifacts/TestCreateCluster_PreTeardownClusterDump/namespaces/e2e-clusters-jj2bs-example-l8xbh/apps/deployments/dns-operator.yaml

I see that the code for other components parameterizes `p.DeploymentConfig.SetDefaultSecurityContext` and sets it based on `r.SetDefaultSecurityContext` where `r` is the `HostedControlPlaneReconciler`.  I'll try doing that.

Reply:
(Sorry, you did specifically mention `r.SetDefaultSecurityContext` in your original comment.  I failed to realize that `r` was referring to `reconcileDNSOperator`'s receiver.)